File: src/my_robot/my_robot/square_driver.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped

class SquareDriver(Node):
    def __init__(self):
        super().__init__('square_driver')
        
        # Sync with Gazebo simulation clock
        self.set_parameters([rclpy.parameter.Parameter('use_sim_time', rclpy.Parameter.Type.BOOL, True)])
        self.publisher_ = self.create_publisher(TwistStamped, '/cmd_vel_out', 10)
        
        # 10Hz loop frequency (0.1 seconds per step)
        self.timer = self.create_timer(0.1, self.timer_callback)
        
        self.counter = 0
        self.side = 0
        
        # States: 
        # 0: Setup-Forward(3m) | 1: Setup-Turn | 2: Setup-Forward(3m) | 3: Setup-Turn
        # 4: Loop-Forward(6m)  | 5: Loop-Turn  | 6: Finished
        self.state = 0 
        self.get_logger().info('Starting outer perimeter mapping routing...')

    def make_twist(self, linear_x=0.0, angular_z=0.0):
        msg = TwistStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'base_link'
        msg.twist.linear.x = linear_x
        msg.twist.angular.z = angular_z
        return msg

    def timer_callback(self):
        # PHASE 1: NAVIGATING TO THE PERIMETER
        if self.state == 0:  # Move 3.0m out from center
            if self.counter < 150:
                self.publisher_.publish(self.make_twist(linear_x=0.2))
                self.counter += 1
            else:
                self.state, self.counter = 1, 0

        elif self.state == 1:  # Turn 90 deg
            if self.counter < 39:
                self.publisher_.publish(self.make_twist(angular_z=0.4))
                self.counter += 1
            else:
                self.state, self.counter = 2, 0

        elif self.state == 2:  # Move 3.0m to reach the first corner link
            if self.counter < 150:
                self.publisher_.publish(self.make_twist(linear_x=0.2))
                self.counter += 1
            else:
                self.state, self.counter = 3, 0

        elif self.state == 3:  # Turn 90 deg to align with wall perimeter
            if self.counter < 39:
                self.publisher_.publish(self.make_twist(angular_z=0.4))
                self.counter += 1
            else:
                self.state, self.counter = 4, 0
                self.get_logger().info('Perimeter reached. Beginning 6m x 6m mapping loop.')

        # PHASE 2: RUNNING THE PERIMETER SQUARE
        elif self.state == 4:  # Drive 6.0m straight parallel to wall (2m clearance)
            if self.counter < 300:
                self.publisher_.publish(self.make_twist(linear_x=0.2))
                self.counter += 1
            else:
                self.state, self.counter = 5, 0

        elif self.state == 5:  # Turn 90 deg at corner
            if self.counter < 39:
                self.publisher_.publish(self.make_twist(angular_z=0.4))
                self.counter += 1
            else:
                self.side += 1
                self.counter = 0
                if self.side < 4:
                    self.state = 4  # Next side
                else:
                    self.state = 6  # Loop complete
                    self.get_logger().info('Map complete! Standing still for Map Saver.')

        elif self.state == 6:  # Final stop
            self.publisher_.publish(self.make_twist(linear_x=0.0, angular_z=0.0))

def main(args=None):
    rclpy.init(args=args)
    node = SquareDriver()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/my_robot/src/square_driver.cpp
#include <memory>

#include "geometry_msgs/msg/twist_stamped.hpp"
#include "rclcpp/rclcpp.hpp"

class SquareDriver : public rclcpp::Node
{
public:
  SquareDriver()
  : Node("square_driver", rclcpp::NodeOptions().parameter_overrides({rclcpp::Parameter("use_sim_time", true)})), counter_(0), side_(0), state_(0)
  {

    publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>("/cmd_vel_out", 10);
    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(100),
      std::bind(&SquareDriver::timer_callback, this));

    RCLCPP_INFO(this->get_logger(), "Starting outer perimeter mapping routing...");
  }

private:
  geometry_msgs::msg::TwistStamped make_twist(double linear_x = 0.0, double angular_z = 0.0)
  {
    geometry_msgs::msg::TwistStamped msg;
    msg.header.stamp = this->get_clock()->now();
    msg.header.frame_id = "base_link";
    msg.twist.linear.x = linear_x;
    msg.twist.angular.z = angular_z;
    return msg;
  }

  void timer_callback()
  {
    if (state_ == 0) {
      if (counter_ < 150) {
        publisher_->publish(make_twist(0.2, 0.0));
        counter_++;
      } else {
        state_ = 1;
        counter_ = 0;
      }
    } else if (state_ == 1) {
      if (counter_ < 39) {
        publisher_->publish(make_twist(0.0, 0.4));
        counter_++;
      } else {
        state_ = 2;
        counter_ = 0;
      }
    } else if (state_ == 2) {
      if (counter_ < 150) {
        publisher_->publish(make_twist(0.2, 0.0));
        counter_++;
      } else {
        state_ = 3;
        counter_ = 0;
      }
    } else if (state_ == 3) {
      if (counter_ < 39) {
        publisher_->publish(make_twist(0.0, 0.4));
        counter_++;
      } else {
        state_ = 4;
        counter_ = 0;
        RCLCPP_INFO(this->get_logger(), "Perimeter reached. Beginning 6m x 6m mapping loop.");
      }
    } else if (state_ == 4) {
      if (counter_ < 300) {
        publisher_->publish(make_twist(0.2, 0.0));
        counter_++;
      } else {
        state_ = 5;
        counter_ = 0;
      }
    } else if (state_ == 5) {
      if (counter_ < 39) {
        publisher_->publish(make_twist(0.0, 0.4));
        counter_++;
      } else {
        side_++;
        counter_ = 0;
        if (side_ < 4) {
          state_ = 4;
        } else {
          state_ = 6;
          RCLCPP_INFO(this->get_logger(), "Map complete! Standing still for Map Saver.");
        }
      }
    } else if (state_ == 6) {
      publisher_->publish(make_twist(0.0, 0.0));
    }
  }

  rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  int counter_;
  int side_;
  int state_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SquareDriver>());
  rclcpp::shutdown();
  return 0;
}
